### PR TITLE
[RPC] Transaction details in getblock

### PR DESCRIPTION
**Pull request by achow101**
Adds an optional parameter extraVerbose to getblock to have transaction details displayed in a getblock call.

### CODE REVIEW

**Comment by dcousens** [comment]
NACK,  is two RPC calls really so bad?

**Comment by achow101** [comment]
@dcousens yes, two RPC calls is really bad, especially when you want the details of all the transactions in a block. Then you end up running getrawtransaction thousands of times. Also, AFAICT, this doesn't require the txindex in order to get the transactions, unlike getrawtransaction.

**Comment by dcousens** [comment]
`getrawtransaction`?
In a batched RPC call,  the overhead is literally just 1RTT more.
Not N,  1.

**Comment by achow101** [comment]
> In a batched RPC call, the overhead is literally just 1RTT more.
Not N (aka, thousands), 1.

How?

> edit: In fact, adding that option would be significantly more useful than this RTT optimisation IMHO.

Well the functionality to this was already in the code. The BlockToJson method took a parameter for txDetails which was by default false. This just lets you set that to true.

**Comment by dcousens** [comment]
http://www.jsonrpc.org/specification#batch

What language are you using? Your RPC library should be able to handle this quite easily.

**Comment by achow101** [comment]
Huh. Didn't know that. I've been using bash with either curl or bitcoin-cli and python.

**Comment by laanwj** [comment]
Saving roundtrip time is not a valid reason to add a RPC.
>  Also, AFAICT, this doesn't require the txindex in order to get the transactions, unlike getrawtransaction.

This is the only important fact here: you *cannot* get this information any other way. getrawtransaction won't get you transactions in blocks, at least without tx index.

**Comment by dcousens** [comment]
@laanwj would it maybe be better to add functionality via getrawtransaction w/ a block id?

**Comment by laanwj** [comment]
> would it maybe be better to add functionality via getrawtransaction w/ a block id? (perhaps in addition to this)

Well yes a "gather transaction [X,...] from prespecified block Y" RPC call could be useful in some rare cases. But on the other hand it'll still have to read the entire block from disk and deserialize it. It's terribly inefficient already.
So if it decoded the entire block why not report details for the entire block?

**Comment by dcousens** [comment]
@laanwj indeed!  Forgot about the whole deserialization.

**Comment by luke-jr** [comment]
Concept ACK for all the reasons @laanwj already discussed.

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -689,11 +689,12 @@ UniValue getblockheader(const UniValue& params, bool fHelp)
 
 UniValue getblock(const UniValue& params, bool fHelp)
 {
-    if (fHelp || params.size() < 1 || params.size() > 2)
+    if (fHelp || params.size() < 1 || params.size() > 3)
         throw runtime_error(
-            "getblock \"hash\" ( verbose )\n"
+            "getblock \"hash\" ( verbose ) ( extraVerbose )\n"
```
Comment (luke-jr) [inline src/rpc/blockchain.cpp]: Double boolean here seems ugly. Maybe allow verbose to be boolean or a number 0-2 (with 2 being extraVerbose)?

Comment (laanwj) [inline src/rpc/blockchain.cpp]: It is ugly, but overloading on type in JSON which is essentially dynamically typed is also ugly.

Named parameters as implemented in #8811 would make this more bearable.

Comment (laanwj) [inline src/rpc/blockchain.cpp]: I do agree that for a new interface, a scale for verbosity would have made more sense instead of a boolean.

**Comment by laanwj** [comment]
I'm starting to realize that @luke-jr's idea to accept 0-2 isn't such a bad idea after all. Just rename the argument  'verbosityLevel'. It is easier to use and understand from a user perspective than two booleans (we regularly get confused there in the tests ourselves). Also it'd remove having to deal with the redundant and nonsensical combo `verbose=false extraVerbose=true`.

For consistency we should do the same in `getrawtransaction`. This method currently accepts a `verbose` argument that can be `0` or `1` but not `false` or `true`. The mapping `false:0,true:1` should be added there (not in this pull though).

**Comment by achow101** [comment]
How should it document in the help message that the old behavior is still accepted?

**Comment by achow101** [comment]
Rebased and made verbose an int from 0-2

**Comment by luke-jr** [comment]
@achow101 The old behaviour should be considered deprecated, and therefore not documented.

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -691,13 +691,17 @@ UniValue getblock(const JSONRPCRequest& request)
...
+            "getblock \"hash\" ( verbose ) \n"
+            "\nIf verbose is 0, returns a string that is serialized, hex-encoded data for block 'hash'.\n"
+            "If verbose is 1, returns an Object with information about block <hash>.\n"
+            "If verbose is 2, returns an Object with information about block <hash> and information about each transaction. \n"
+            "The previous behavior of verbose being a boolean is still maintained. False will have the behavior of 0 and True will have the behavior of 1.\n"
             "\nArguments:\n"
             "1. \"hash\"          (string, required) The block hash\n"
-            "2. verbose           (boolean, optional, default=true) true for a json object, false for the hex encoded data\n"
-            "\nResult (for verbose = true):\n"
+            "2. verbose           (boolean, optional, default=1) 0 for hex encoded data, 1 for a json object, and 2 for json object with transaction data\n"
```
Comment (luke-jr) [inline src/rpc/blockchain.cpp]: Should be renamed to verbosity.

**Comment by luke-jr** [comment]
Also, fix the commit description...

**Comment by achow101** [comment]
@luke-jr done

**Comment by achow101** [comment]
merge please?

**Comment by achow101** [comment]
merge please?

**Review by luke-jr — APPROVED** [review]
utACK

**Comment by luke-jr** [comment]
(Needs rebase)

**Comment by achow101** [comment]
rebased

**Review by luke-jr — APPROVED** [review]

**Comment by pstratem** [comment]
Is this compatible with callers of getblock which send true instead of 0/1 ?

**Comment by achow101** [comment]
@pstratem Yes. The old style of calling is maintained for compatibility.

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -721,8 +724,72 @@ UniValue getblock(const JSONRPCRequest& request)
...
+            "  \"hash\" : \"hash\",     (string) the block hash (same as provided)\n"
+            "  \"confirmations\" : n,   (numeric) The number of confirmations, or -1 if the block is not on the main chain\n"
+            "  \"size\" : n,            (numeric) The block size\n"
+            "  \"strippedsize\" : n,    (numeric) The block size excluding witness data\n"
+            "  \"weight\" : n           (numeric) The block weight (BIP 141)\n"
+            "  \"height\" : n,          (numeric) The block height or index\n"
+            "  \"version\" : n,         (numeric) The block version\n"
+            "  \"versionHex\" : \"00000000\", (string) The block version formatted in hexadecimal\n"
+            "  \"merkleroot\" : \"xxxx\", (string) The merkle root\n"
+            "  \"tx\" : [               (array of Objects) The transactions\n"
```
Comment (jnewbery) [inline src/rpc/blockchain.cpp]: I'd suggest not including this in the help output, and just having something like:

```c
           "  \"tx\" : [               (array of Objects) The transactions. Format is the same as for getrawtransaction call\n"
           "         ,...\n"
           "  ],\n"
```

If the output format of `TxToJSON()` ever changes, I'm sure that changing this help text will be forgotten!

**Comment by sipa** [comment]
Needs rebase.

**Comment by achow101** [comment]
rebased. also added @jnewbery's suggestion with the help text.

**Review by kallewoof — COMMENTED** [review]
utACK b779f3093d7b68ee09305a5294c48825dce8ae1f

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -716,8 +719,29 @@ UniValue getblock(const JSONRPCRequest& request)
             "  \"previousblockhash\" : \"hash\",  (string) The hash of the previous block\n"
             "  \"nextblockhash\" : \"hash\"       (string) The hash of the next block\n"
             "}\n"
-            "\nResult (for verbose=false):\n"
-            "\"data\"             (string) A string that is serialized, hex-encoded data for block 'hash'.\n"
+            "\nResult (for verbosity = 2):\n"
```
Comment (paveljanik) [inline src/rpc/blockchain.cpp]: What about mentioning only the differences to verbosity=1 instead?

Comment (achow101) [inline src/rpc/blockchain.cpp]: something like this?

```
{
..., Same as for verbosity=1
"tx" : [           (array of Objects) The transactions in the format of the getrawtransaction RPC.
         , ...
],
... Same as for verbosity=1
}
```

Comment (paveljanik) [inline src/rpc/blockchain.cpp]: Yup, or maybe even better:

```
Additional output (for verbosity=1):...
```

**Comment by achow101** [comment]
@paveljanik I made the change.

**Comment by laanwj** [comment]
Tested ACK e3c9f2d

**Pull request merged**